darts/utils/model_selection.py

```diff
@@ -0,0 +1,194 @@
+                 axis: Optional[int] = 0,
+                 input_size: Optional[int] = 0,
+                 horizon: Optional[int] = 0,
+                 vertical_split_type: Optional[str] = 'simple'):
+
+        if type not in ['train', 'test']:
+            raise AttributeError('Value for type parameter should be either `train` or `test`')
+        self.type = type
+
+        if not data:
+            raise AttributeError('The `data` parameter cannot be empty.')
+        self.data = data
+
+        self.test_size = test_size
+        self.axis = axis
+
+        if axis == 1 and vertical_split_type == 'model-aware' and (horizon == 0 or input_size == 0):
+                raise AttributeError("You need to provide non-zero `horizon` and `n` parameters when axis=1")
+
+        self.input_size = input_size
+        self.horizon = horizon
+
+        if vertical_split_type not in ['simple', 'model-aware']:
+            raise AttributeError('`vertical_split_type` can be eiter `simple` or `model-aware`.')
+        else:
+            self.vertical_split_type = vertical_split_type
+
+    def _get_horizontal_split_index(self):
+        if 0 < self.test_size < 1:
+            return int(len(self.data) * (1 - self.test_size))
+        else:
+            return len(self.data) - self.test_size
+
+    def _get_vertical_split_indices(self, ts_length):
+
+        if self.vertical_split_type == 'simple':
+            if 0 < self.test_size < 1:
+                test_size = int(ts_length * self.test_size)
+            else:
+                test_size = self.test_size
+
+            test_start_index = ts_length - test_size
+            train_end_index = test_start_index
+
+            if test_start_index < 0:
+                raise AttributeError("`test_size` is bigger then timeseries length")
+
+        else: # model-aware split
+            train_end_index = ts_length - self.horizon
+
+            if 0 < self.test_size < 1:
+                test_size = int((ts_length - self.horizon) * (self.test_size))
+            else:
+                test_size = self.test_size
+
+            if train_end_index < 0:
+                train_end_index = 0
+
+            if train_end_index < self.input_size:
+                raise AttributeError("Training timeseries is of 0 size")
+
+            test_start_index = ts_length - self.horizon - self.input_size - test_size - 1
+
+            if test_start_index < 0:
+                raise AttributeError("Not enough timesteps to create testset")
+
+        return train_end_index, test_start_index
+
+    def __getitem__(self, i: int) -> TimeSeries:
+        if self.axis == 0:
+            split_index = self._get_horizontal_split_index()
+            if self.type == 'train':
+                if i >= split_index:
+                    raise IndexError('Exceeded the size of the sequence.')
+                return self.data[i]
+            else:
+                if i + split_index > len(self.data):
```

Comment:
It can't be `>=` since we are comparing 0-based indicies to length of the array. To satisfy `>=` it would have to be `i + split_index >= len(self.data) - 1`, because `len(self.data) - 1` gives the last index.

Reply:
there are tests that test it, btw: `test_horiz_split_overindexing_test_set()`, `test_horiz_split_last_index_test_set()`

Reply:
Yes and I guess the tests are not catching it because in the `=` case line 99 below will also anyway return an `IndexError`, which is what your test expects.

Reply:
Oooomph. I thought that `assertRaise()` parameter `msg=` checks whether message from the exception is correct. Turns out that you have to use `assertRaiseRegex()` to do that.